stubs/docutils/docutils/nodes.pyi

```diff
@@ -1,15 +1,75 @@
+import xml.dom.minidom
 from _typeshed import Self
-from collections.abc import Iterable
-from typing import Any, overload
+from collections.abc import Iterable, Sequence
+from typing import Any, Callable, Generator, Protocol, TypeVar, overload
+from typing_extensions import Literal
 
 from docutils.transforms import Transformer
 
+class NodeVisitor:
+    def __init__(self, document: document): ...
+    def __getattr__(self, __name: str) -> Any: ...  # incomplete
+
+_N = TypeVar("_N", bound=Node)
+
+class _DomModule(Protocol):
+    Document: type[xml.dom.minidom.Document]
+
 class Node:
+    children: Sequence[Node]
     parent: Node | None
     source: str | None
     line: int | None
     document: document | None
-    def __getattr__(self, __name: str) -> Any: ...  # incomplete
+    def __bool__(self) -> Literal[True]: ...
+    def asdom(self, dom: _DomModule | None = ...) -> xml.dom.minidom.Element: ...
+    def pformat(self, indent: str = ..., level: int = ...) -> str: ...
+    def copy(self: Self) -> Self: ...
+    def deepcopy(self: Self) -> Self: ...
+    def astext(self) -> str: ...
```

Comment:
As these are abstract, they should be annotated using `@abc.abstractmethod`, even if they are not annotated in the implementation. This also means that they needs to be copied to `Text` (as non-abstract).

Reply:
What is the benefit of adding `@abc.abstractmethod`?

Reply:
Documentation and tooling can - in theory - flag improper use of these methods.

Reply:
What would improper use be?

Reply:
It's a bit of a weird example, given that it's a deprecated module, but it's a good example: `asynchat.async_chat` doesn't have any abstract methods at runtime, and doesn't have `ABCMeta` as its metaclass. But its docstring nonetheless states that it is an "abstract class", and so we decorate the abstract methods with `@abstractmethod` in typeshed. As a result, mypy will raise an error if you try to directly instantiate the class (the proper way to use the class is to create a concrete subclass that overrides the abstract methods, and then instantiate an instance of the subclass).

- Runtime: https://github.com/python/cpython/blob/4558af5a8f8e56a9b0dc11f6e834c47e0fd05f9e/Lib/asynchat.py#L60-L62
- typeshed stub: https://github.com/python/typeshed/blob/64133df6eb3f617c2e2b055a66943fc4503d82ce/stdlib/asynchat.pyi#L9-L16
- mypy treatment of this: https://mypy-play.net/?mypy=latest&python=3.10&gist=947e63ddae68c9ab2269ac7ce42da22d

I'm not aware whether pyright also has this feature, but the point is that it _could_ implement this feature if it wanted to.

Reply:
Thanks for the thorough explanation and the additional example! I asked Eric about adding that check to pyright but declined since the check is expensive.

Anyway I added the `@abstactmedhod` decorators but now mypy.stubtest is failing:

> If it is meant to be abstract, add 'abc.ABCMeta' as an explicit metaclass

Reply:
Thanks. `pformat` and `astext` should also be abstract. Stubtest is failing, because these four methods now need to be copied to `Element` and `document` (non-abstract). Sorry for the hassle.